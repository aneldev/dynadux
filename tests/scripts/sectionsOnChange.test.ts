import "jest";
import {
  ICreateStoreAPI,
  createStore
} from "../../src/create/createStore";

// User Section store

interface IUserInfoState {
  logged: boolean;
  name: string;
  avatar: string;
  loggedAt: string;
}

enum EUserActions {
  LOGIN = "LOGIN",                  // payload: ILOGIN_payload
  LOGOUT = "LOGOUT",
  UPDATE_AVATAR = "UPDATE_AVATAR",  // payload: string: new url
}

interface ILOGIN_payload {
  name: string;
  avatar: string;
}

const createUserInfoSection = (store: ICreateStoreAPI) => {
  const changes: { state: IUserInfoState; action: string; payload: any; }[] = [];
  const section = store.createSection<IUserInfoState>({
    section: 'userSection',
    onChange: (state, action, payload) => changes.push({state, action, payload}),
    initialState: {
      logged: false,
      name: '',
      avatar: '',
      loggedAt: '',
    },
    reducers: {
      [EUserActions.LOGIN]: ({payload}) => {
        const {
          name,
          avatar,
        }: ILOGIN_payload = payload;
        return {
          logged: true,
          name,
          avatar,
          loggedAt: "today",
        };
      },
      [EUserActions.LOGOUT]: ({state: {logged}}) => {
        if (!logged) return; // Exit. No need to change anything.
        return {
          logged: false,
          name: '',
          avatar: '',
          loggedAt: "today",
        };
      },
      [EUserActions.UPDATE_AVATAR]: ({payload: avatar}) => ({avatar}),
    },
  });

  return {
    get state() {
      return section.state;
    },
    get changes() {
      return changes;
    },
    actions: {
      login: (name: string, avatar: string) => section.dispatch<ILOGIN_payload>(EUserActions.LOGIN, {name, avatar}),
      logout: () => section.dispatch(EUserActions.LOGOUT),
      updateAvatar: (avatar: string) => section.dispatch<string>(EUserActions.UPDATE_AVATAR, avatar),
    }
  };
};

// Todos Section store

interface ITodosManagementState {
  todos: ITodo[];
  lastAddedTodo: string;
}

interface ITodo {
  id: number;
  label: string;
  done: boolean;
}

enum ETodosActions {
  ADD_TODO = "ADD_TODO",            // payload: IADD_TODO_payload
  REMOVE_TODO = "REMOVE_TODO",      // payload: number: the id of the todo
  COMPLETE_TODO = "COMPLETE_TODO",  // payload: number: the id of the todo
}

interface IADD_TODO_payload {
  id: number;
  label: string;
}

const createTodosSection = (store: ICreateStoreAPI) => {
  const changes: { state: ITodosManagementState; action: string; payload: any; }[] = [];
  const section = store.createSection<ITodosManagementState>({
    section: 'todosSection',
    onChange: (state, action, payload) => changes.push({state, action, payload}),
    initialState: {
      todos: [],
      lastAddedTodo: '',
    },
    reducers: {
      [ETodosActions.ADD_TODO]: ({state: {todos}, payload}) => {
        const {id, label}: IADD_TODO_payload = payload;
        return {
          todos: todos.concat({id, label, done: false}),
          lastAddedTodo: "today",
        };
      },
      [ETodosActions.REMOVE_TODO]: ({state: {todos}, payload: id}) => {
        return {
          todos: todos.filter(todo => todo.id !== id),
        };
      },
      [ETodosActions.COMPLETE_TODO]: ({state: {todos}, payload: id}) => {
        return {
          todos: todos.map(todo => {
            if (todo.id === id) return ({...todo, done: true});
            return todo;
          }),
        };
      },
    },
  });

  return {
    get state() {
      return section.state;
    },
    get changes() {
      return changes;
    },
    actions: {
      addTodo: (id: number, label: string) => section.dispatch<IADD_TODO_payload>(ETodosActions.ADD_TODO, {id, label}),
      removeTodo: (id: number) => section.dispatch<number>(ETodosActions.REMOVE_TODO, id),
      completeTodo: (id: number) => section.dispatch<number>(ETodosActions.COMPLETE_TODO, id),
    }
  };

};

// App store

interface IAppState {
  clientId: string;
  userSection: IUserInfoState;
  todosSection: ITodosManagementState;
}

enum EAppActions {
  SET_CLIENT_ID = "SET_CLIENT_ID"
}

const createAppStore = () => {
  const changes: { state: IAppState; action: string; payload?: any }[] = [];
  const store = createStore<IAppState>({
    initialState: {
      clientId: '',
      userSection: null as any,
      todosSection: null as any,
    },
    reducers: {
      [EAppActions.SET_CLIENT_ID]: ({payload: clientId}) => {
        return {clientId};
      },
    },
    onChange: (state, action, payload) => changes.push({state, action, payload}),
  });

  return {
    get state() {
      return store.state;
    },
    get changes() {
      return changes;
    },
    setClientId: (clientId: string): void => store.dispatch<string>(EAppActions.SET_CLIENT_ID, clientId),
    user: createUserInfoSection(store),
    todos: createTodosSection(store),
  };
};


describe('Dynadux', () => {
  test('Working with Sections', () => {
    const store = createAppStore();

    expect(store.state).toMatchSnapshot('Initial state');

    store.user.actions.login('John', 'https://www.anel.co/user/928457123/profile-01.png');
    store.todos.actions.addTodo(101, 'Before work beers');
    store.todos.actions.addTodo(102, 'After work beers');

    store.setClientId('client-001');

    expect(store.state).toMatchSnapshot('First todos');

    store.todos.actions.removeTodo(101);

    expect(store.state).toMatchSnapshot('After remove of 101 todo');

    store.todos.actions.completeTodo(102);

    expect(store.state).toMatchSnapshot('After complete 102');

    store.setClientId('client-002');

    store.user.actions.updateAvatar('https://www.anel.co/user/928457123/profile-02.png');

    expect(store.state).toMatchSnapshot('After avatar update');

    store.setClientId('client-003');

    expect(store.changes).toMatchSnapshot('STORE-CHANGES');
    expect(store.user.changes).toMatchSnapshot('USER-SECTION-CHANGES');
    expect(store.todos.changes).toMatchSnapshot('TODO-SECTION-CHANGES');
  });
});
