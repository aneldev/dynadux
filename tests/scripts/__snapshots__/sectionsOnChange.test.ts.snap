// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynadux Working with Sections: After avatar update 1`] = `
Object {
  "clientId": "client-002",
  "todosSection": Object {
    "lastAddedTodo": "today",
    "todos": Array [
      Object {
        "done": true,
        "id": 102,
        "label": "After work beers",
      },
    ],
  },
  "userSection": Object {
    "avatar": "https://www.anel.co/user/928457123/profile-02.png",
    "logged": true,
    "loggedAt": "today",
    "name": "John",
  },
}
`;

exports[`Dynadux Working with Sections: After complete 102 1`] = `
Object {
  "clientId": "client-001",
  "todosSection": Object {
    "lastAddedTodo": "today",
    "todos": Array [
      Object {
        "done": true,
        "id": 102,
        "label": "After work beers",
      },
    ],
  },
  "userSection": Object {
    "avatar": "https://www.anel.co/user/928457123/profile-01.png",
    "logged": true,
    "loggedAt": "today",
    "name": "John",
  },
}
`;

exports[`Dynadux Working with Sections: After remove of 101 todo 1`] = `
Object {
  "clientId": "client-001",
  "todosSection": Object {
    "lastAddedTodo": "today",
    "todos": Array [
      Object {
        "done": false,
        "id": 102,
        "label": "After work beers",
      },
    ],
  },
  "userSection": Object {
    "avatar": "https://www.anel.co/user/928457123/profile-01.png",
    "logged": true,
    "loggedAt": "today",
    "name": "John",
  },
}
`;

exports[`Dynadux Working with Sections: First todos 1`] = `
Object {
  "clientId": "client-001",
  "todosSection": Object {
    "lastAddedTodo": "today",
    "todos": Array [
      Object {
        "done": false,
        "id": 101,
        "label": "Before work beers",
      },
      Object {
        "done": false,
        "id": 102,
        "label": "After work beers",
      },
    ],
  },
  "userSection": Object {
    "avatar": "https://www.anel.co/user/928457123/profile-01.png",
    "logged": true,
    "loggedAt": "today",
    "name": "John",
  },
}
`;

exports[`Dynadux Working with Sections: Initial state 1`] = `
Object {
  "clientId": "",
  "todosSection": Object {
    "lastAddedTodo": "",
    "todos": Array [],
  },
  "userSection": Object {
    "avatar": "",
    "logged": false,
    "loggedAt": "",
    "name": "",
  },
}
`;

exports[`Dynadux Working with Sections: STORE-CHANGES 1`] = `
Array [
  Object {
    "action": "LOGIN",
    "payload": Object {
      "avatar": "https://www.anel.co/user/928457123/profile-01.png",
      "name": "John",
    },
    "state": Object {
      "clientId": "",
      "todosSection": Object {
        "lastAddedTodo": "",
        "todos": Array [],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "ADD_TODO",
    "payload": Object {
      "id": 101,
      "label": "Before work beers",
    },
    "state": Object {
      "clientId": "",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": false,
            "id": 101,
            "label": "Before work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "ADD_TODO",
    "payload": Object {
      "id": 102,
      "label": "After work beers",
    },
    "state": Object {
      "clientId": "",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": false,
            "id": 101,
            "label": "Before work beers",
          },
          Object {
            "done": false,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "SET_CLIENT_ID",
    "payload": "client-001",
    "state": Object {
      "clientId": "client-001",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": false,
            "id": 101,
            "label": "Before work beers",
          },
          Object {
            "done": false,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "REMOVE_TODO",
    "payload": 101,
    "state": Object {
      "clientId": "client-001",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": false,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "COMPLETE_TODO",
    "payload": 102,
    "state": Object {
      "clientId": "client-001",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": true,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "SET_CLIENT_ID",
    "payload": "client-002",
    "state": Object {
      "clientId": "client-002",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": true,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-01.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "UPDATE_AVATAR",
    "payload": "https://www.anel.co/user/928457123/profile-02.png",
    "state": Object {
      "clientId": "client-002",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": true,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-02.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
  Object {
    "action": "SET_CLIENT_ID",
    "payload": "client-003",
    "state": Object {
      "clientId": "client-003",
      "todosSection": Object {
        "lastAddedTodo": "today",
        "todos": Array [
          Object {
            "done": true,
            "id": 102,
            "label": "After work beers",
          },
        ],
      },
      "userSection": Object {
        "avatar": "https://www.anel.co/user/928457123/profile-02.png",
        "logged": true,
        "loggedAt": "today",
        "name": "John",
      },
    },
  },
]
`;

exports[`Dynadux Working with Sections: TODO-SECTION-CHANGES 1`] = `
Array [
  Object {
    "action": "ADD_TODO",
    "payload": Object {
      "id": 101,
      "label": "Before work beers",
    },
    "state": Object {
      "lastAddedTodo": "today",
      "todos": Array [
        Object {
          "done": false,
          "id": 101,
          "label": "Before work beers",
        },
      ],
    },
  },
  Object {
    "action": "ADD_TODO",
    "payload": Object {
      "id": 102,
      "label": "After work beers",
    },
    "state": Object {
      "lastAddedTodo": "today",
      "todos": Array [
        Object {
          "done": false,
          "id": 101,
          "label": "Before work beers",
        },
        Object {
          "done": false,
          "id": 102,
          "label": "After work beers",
        },
      ],
    },
  },
  Object {
    "action": "REMOVE_TODO",
    "payload": 101,
    "state": Object {
      "lastAddedTodo": "today",
      "todos": Array [
        Object {
          "done": false,
          "id": 102,
          "label": "After work beers",
        },
      ],
    },
  },
  Object {
    "action": "COMPLETE_TODO",
    "payload": 102,
    "state": Object {
      "lastAddedTodo": "today",
      "todos": Array [
        Object {
          "done": true,
          "id": 102,
          "label": "After work beers",
        },
      ],
    },
  },
]
`;

exports[`Dynadux Working with Sections: USER-SECTION-CHANGES 1`] = `
Array [
  Object {
    "action": "LOGIN",
    "payload": Object {
      "avatar": "https://www.anel.co/user/928457123/profile-01.png",
      "name": "John",
    },
    "state": Object {
      "avatar": "https://www.anel.co/user/928457123/profile-01.png",
      "logged": true,
      "loggedAt": "today",
      "name": "John",
    },
  },
  Object {
    "action": "UPDATE_AVATAR",
    "payload": "https://www.anel.co/user/928457123/profile-02.png",
    "state": Object {
      "avatar": "https://www.anel.co/user/928457123/profile-02.png",
      "logged": true,
      "loggedAt": "today",
      "name": "John",
    },
  },
]
`;
